implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java

```diff
@@ -269,12 +274,26 @@ public SmallRyeHealth getHealthGroups() {
     }
 
     private Uni<SmallRyeHealth> getHealthAsync(Uni<SmallRyeHealth> cachedHealth, HealthType... types) {
-        if (additionalListsChanged(types) || additionalListsChanged || cachedHealth == null) {
-            additionalListsChanged = false;
-            cachedHealth = computeHealth(types);
+        if (contextPropagated) {
+            recreateCheckUnis();
+            return computeHealth(types);
+        } else {
+            if (additionalListsChanged(types) || additionalListsChanged || cachedHealth == null) {
+                additionalListsChanged = false;
+                cachedHealth = computeHealth(types);
+            }
+
+            return cachedHealth;
         }
+    }
+
+    private void recreateCheckUnis() {
+        livenessUnis.clear();
```

Comment:
@xstefank  are you sure that it is a good idea to mutate the global state that could be accessed concurrently?  
 it is better to create new variant of  computeHealth method without side effects

Reply:
@robotmrv maybe. In my integrations, I wouldn't expect concurrent calls to this method. However, better to be safe than sorry. I'll look into it.